packages/contracts-bedrock/test/setup/Upgrade.s.sol

```diff
@@ -0,0 +1,104 @@
+
+        // Bridge contracts
+        address optimismPortal = vm.parseTomlAddress(opToml, ".addresses.OptimismPortalProxy");
+        save("OptimismPortalProxy", optimismPortal);
+        save(
+            "OptimismPortal", address(uint160(uint256(vm.load(optimismPortal, Constants.PROXY_IMPLEMENTATION_ADDRESS))))
+        );
+        save("OptimismPortal2", optimismPortal);
+        address addressManager = vm.parseTomlAddress(opToml, ".addresses.AddressManager");
+        save("AddressManager", addressManager);
+        save("L1CrossDomainMessenger", IAddressManager(addressManager).getAddress("OVM_L1CrossDomainMessenger"));
+        save("L1CrossDomainMessengerProxy", vm.parseTomlAddress(opToml, ".addresses.L1CrossDomainMessengerProxy"));
+        saveProxyAndImpl("OptimismMintableERC20Factory", opToml, ".addresses.OptimismMintableERC20FactoryProxy");
+        saveProxyAndImpl("L1StandardBridge", opToml, ".addresses.L1StandardBridgeProxy");
+        saveProxyAndImpl("L1ERC721Bridge", opToml, ".addresses.L1ERC721BridgeProxy");
+
+        // Fault proof proxied contracts
+        saveProxyAndImpl("AnchorStateRegistry", opToml, ".addresses.AnchorStateRegistryProxy");
+        saveProxyAndImpl("DisputeGameFactory", opToml, ".addresses.DisputeGameFactoryProxy");
+        saveProxyAndImpl("DelayedWETH", opToml, ".addresses.DelayedWETHProxy");
+
+        // Fault proof non-proxied contracts
+        save("PreimageOracle", vm.parseTomlAddress(opToml, ".addresses.PreimageOracle"));
+        save("Mips", vm.parseTomlAddress(opToml, ".addresses.MIPS"));
+        IDisputeGameFactory disputeGameFactory = IDisputeGameFactory(mustGetAddress("DisputeGameFactoryProxy"));
+        save("FaultDisputeGame", vm.parseTomlAddress(opToml, ".addresses.FaultDisputeGame"));
+        // The PermissionedDisputeGame and PermissionedDelayedWETHProxy are not listed in the registry for OP, so we
+        // look it up onchain
+        IFaultDisputeGame permissionedDisputeGame =
+            IFaultDisputeGame(address(disputeGameFactory.gameImpls(GameTypes.PERMISSIONED_CANNON)));
+        save("PermissionedDisputeGame", address(permissionedDisputeGame));
+        save("PermissionedDelayedWETHProxy", address(permissionedDisputeGame.weth()));
+    }
+
+    /// @notice Saves the proxy and implementation addresses for a contract name
+    /// @param _contractName The name of the contract to save
+    /// @param _tomlPath The path to the superchain config file
+    /// @param _tomlKey The key in the superchain config file to get the proxy address
+    function saveProxyAndImpl(string memory _contractName, string memory _tomlPath, string memory _tomlKey) internal {
+        address proxy = vm.parseTomlAddress(_tomlPath, _tomlKey);
+        save(string.concat(_contractName, "Proxy"), proxy);
+        save(_contractName, address(uint160(uint256(vm.load(proxy, Constants.PROXY_IMPLEMENTATION_ADDRESS)))));
```

Comment:
Should we verify that the implementation is not the zero address? Same for the other `saveProxyAndImpl`

Reply:
sure, done.